packages/cli/src/services/redis/redis-client.service.ts

```diff
@@ -0,0 +1,146 @@
+import { Service } from 'typedi';
+import config from '@/config';
+import { Logger } from '@/Logger';
+import ioRedis from 'ioredis';
+import type { RedisOptions } from 'ioredis';
+import type { RedisClientType } from './RedisServiceBaseClasses';
+
+@Service()
+export class RedisClientService {
+	constructor(private readonly logger: Logger) {}
+
+	createClient(arg: { type: RedisClientType; extraOptions?: RedisOptions }) {
+		return this.clusterNodes().length > 0
```

Comment:
We could keep all these clients in a `WeakSet` and add a `@OnShutdown()` handler to call `.disconnect()` on each of the clients created by this service.

Reply:
`WeakSet` is not iterable - or am I misunderstanding what you mean?

```ts
export class RedisClientService {
	clients = new WeakSet<ioRedis | Cluster>();

	constructor(private readonly logger: Logger) {}

	createClient(arg: { type: RedisClientType; extraOptions?: RedisOptions }) {
		const client =
			this.clusterNodes().length > 0
				? this.createClusterClient(arg)
				: this.createRegularClient(arg);

		this.clients.add(client);

		return client;
	}

	@OnShutdown()
	disconnectClients() {
		for (const client of this.clients) { // Type 'WeakSet<Redis | Cluster>' must have a '[Symbol.iterator]()' method that returns an iterator.ts(2488)
			client.disconnect();
		}
	}
}
```

Reply:
I forgot that `Weakset` isn't `Iterable`. Maybe we can use a `Set` instead.

Reply:
Thank you, please see [67b3d56](https://github.com/n8n-io/n8n/pull/9774/commits/67b3d5647f8bcd4c368f143582b587304d8be2de)